### PR TITLE
Add UnitHook and NoopHook implementations

### PR DESCRIPTION
**Pull request by mcastorina**
### Description:
The UnitHook tracks metrics per unit of a job, and safely makes them available to the caller. It also works for sources that do not support source units.

### Checklist:
* [ ] Tests passing (`make test-community`)?
* [ ] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/usage/install/#local-installation))?

### CODE REVIEW

File: pkg/sources/job_progress_hook.go [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+	// lru.NewWithEvict can only fail if the size is < 0.
+	cache, _ := lru.NewWithEvict(1024, func(key string, value *UnitMetrics) {
+		if value.IsFinished() {
+			return
+		}
+		ctx.Logger().Error(fmt.Errorf("eviction"), "dropping unit metric",
+			"id", key,
+			"metric", value,
+		)
+	})
```
Comment (mcastorina) [inline pkg/sources/job_progress_hook.go:38]: This eviction function is called even when we manually call `Remove`, which is I added a check for `value.IsFinished()`.

However, it's possible for a value to be finished and evicted without ever being consumed. In fact, that's probably more common than evicting an in-progress metric. In this case, we wouldn't see the log.

TODO: Refactor this callback to log for unhandled evictions.

File: go.mod [hunk 2]
```diff
@@ -47,6 +47,7 @@ require (
 	github.com/h2non/filetype v1.1.3
 	github.com/hashicorp/go-retryablehttp v0.7.4
 	github.com/hashicorp/golang-lru v0.5.1
+	github.com/hashicorp/golang-lru/v2 v2.0.7
```
Comment (rosecodym) [inline go.mod:50]: @ahrav i thought we already had an lru cache somewhere? does it just not use this package?

**Review by rosecodym — APPROVED** [review]
is your approach to these metrics documented anywhere? i don't recall anything about them in the original design doc (maybe i'm just forgetting them) and i would like a high level overview of your strategy

File: go.mod (hunk 2 above)
Comment (ahrav) [reply go.mod:50]: https://github.com/trufflesecurity/trufflehog/blob/68f28a0e34f9f95bbfc533595756c6525c7b508f/go.mod#L49C7-L49C7 yep.

Comment (ahrav) [reply go.mod:50]: maybe not v2?

**Comment by mcastorina** [comment]
> is your approach to these metrics documented anywhere? i don't recall anything about them in the original design doc (maybe i'm just forgetting them) and i would like a high level overview of your strategy

No, this approach came about mostly through exploratory coding.

Thanks for the callout, I'll write up a doc!

**Pull request merged**